### ISSUE TITLE
Setting Cache Control

### ISSUE THREAD

**Issue by michaelstivala** [spatie/laravel-medialibrary#132]
Hello Spatie.

I'm trying to solve a particular problem when it comes to setting the browser cache for resources stored on external disks - to elaborate, I upload my media to S3 and then use cloudfront as a CDN to deliver the media as quickly as possible.

The above works as expected using your *excellent* package. The issue was brought to my attention through Google's PageSpeed Insight tool which complains that media hosted on cloudfront doesn't leverage browser caching:

![image](https://cloud.githubusercontent.com/assets/4493561/12145600/776ab8ee-b48f-11e5-86f9-798ecb235aaf.png)

Below are some possible solutions.

*Manually change headers*
Manually log into S3 and set the Cache-Control property, this solves the issue - but is obviously not a viable solution. 

*Pragramatically add headers*
Add the ability to set the Cache-Control header (or other headers) via this package. I tested this quickly by modifying Spatie\MediaLibrary\Filesystem::copyToMediaLibrary to read as follows:

```
public function copyToMediaLibrary($file, Media $media, $conversions = false, $targetFileName = '')
    {
        $destination = $this->getMediaDirectory($media, $conversions).
            ($targetFileName == '' ? pathinfo($file, PATHINFO_BASENAME) : $targetFileName);

        if ($media->getDiskDriverName() === 'local') {
            $this->filesystem
                ->disk($media->disk)
                ->put($destination, fopen($file, 'r+'));
        } else {
            $this->filesystem
                ->disk($media->disk)
                ->getDriver()
                ->put($destination, fopen($file, 'r+'), ['ContentType' => File::getMimeType($file), 'CacheControl' => 'max-age=604800']);
        }
    }
```
(Note the new 'CacheControl' key in the array)

The ability to add the Cache-Control header could be moved to a config to give the developer more control, or potentially be expanded to allow any additional headers to be added for each file.

*Create a custom S3 Adaptor*
I haven't tried this yet, but referencing this [stackoverflow question](http://stackoverflow.com/a/30269860/2426387), the developer could write a custom adaptor that includes default settings for file upload, specifying and defaults that the project requires.

Do you have any thoughts and opinions on the above? I'd love to hear feedback from the creators.

Thanks for listening, and for providing this package.

**Comment by freekmurze** [comment]
Hi Maltese Rockstar,

thank you very much for this quality explanation of the problem. I also appreciate that you suggest some quality solutions. Good job. :+1: 

I'm all for the pragmatic solution. Maybe we could introduce a `extra_headers` key in the config file that contains an array of options that would be merge into the second parameters of the `put`-function. In that way people can set all the extra headers that they want. A sane default could be ['CacheControl' => 'max-age=604800']`.

Thoughts?

**Comment by michaelstivala** [comment]
Sounds good Freek! 

I can work on a PR for this - if there's anything you'd like me to keep in mind while on it to facilitate it getting merged in quicker, let me know.

Thanks! :)

**Comment by freekmurze** [comment]
This is implemented in v3.12.0 of the package.
